File: BOJ/12015.c
#include <stdio.h>
#define MAXNUM 1000000
int main()
{
    int A;
    int Ai[MAXNUM];
    int data[MAXNUM] = {
        0,
    };
    int maxDataIndex = 0;
    scanf("%d", &A);
    for (int i = 0; i < A; i++)
    {
        scanf("%d", &Ai[i]);
    }

    for (int i = 0; i < A; i++)
    {
        int n = Ai[i];
        if (n > data[maxDataIndex])
        {
            maxDataIndex += 1;
            data[maxDataIndex] = n;
        }
        else
        {
            for (int j = 1; j <= maxDataIndex; j++)
            {
                if (n <= data[j])
                {
                    data[j] = n;
                    break;
                }
            }
        }
    }
    printf("%d\n", maxDataIndex);
}

File: BOJ/12015.py
A = int(input())
Ai = list(map(int, input().split()))

data = [0]*1000000
maxDataIndex = 0
for n in Ai:
    if n > data[maxDataIndex]:
        maxDataIndex += 1
        data[maxDataIndex] = n
    else:
        for i in range(1, len(data)):
            if n <= data[i]:
                data[i] = n
                break
# print(data[1:maxDataIndex+1])
print(maxDataIndex)
